File: OverviewInformationTechnology/Program1/C/Section1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct buffer {
    char * buf;  // buffer
    int len; // length
    int cap; // capacity
};

#define BUFFER_EXTEND_SIZE 256

int newString(char * content, int size, struct buffer * result) {
    char * buf = (char *)malloc(size * 2);
    if(buf == NULL) return 0;
    memcpy(buf, content, size);
    result->buf = buf;
    result->len = size;
    result->cap = size * 2;
    return 1;
}

int insertString(struct buffer * me, char * content, int start, int size) {
    if(me->cap - me->len < size) {
        int new_cap = me->cap + BUFFER_EXTEND_SIZE;
        while(new_cap < me->len + size)
            new_cap += BUFFER_EXTEND_SIZE;
        char * new_buf = (char *)realloc(me->buf, new_cap);
        if(new_buf != NULL) {
            me->buf = new_buf;
            me->cap = new_cap;
        } else {
            new_buf = (char *)malloc(new_cap);
            if(new_buf != NULL) {
                memcpy(new_buf, me->buf, me->len);
                free(me->buf);
                me->buf = new_buf;
                me->cap = new_cap;
            } else {
                return 0;
            }
        }
    }
    start = start >= me->len ? me->len : start;
    for(int i = me->len + size - 1; i >= start + size; --i)
        me->buf[i] = me->buf[i-size];
    memcpy(me->buf + start, content, size);
    me->len += size;
    return 1;
}

void removeString(struct buffer * me, int start, int size) {
    if(start >= me->len) return;
    int end = size + start;
    if(end > me->len) end = me->len;
    for(int i = 0; end + i < me->len; ++i) {
        me->buf[i + start] = me->buf[end + i];
    }
    me->len -= (end-start);
}

int isNum(char c) {
    return c >= '0' && c <= '9';
}

int parse(char * buf, int * start, int * end) {
    buf = buf + 1;
    int num_tail_pos = 0;
    int cont = 0;
    for(; buf[num_tail_pos] != '\n'; ++num_tail_pos) {
        if(buf[num_tail_pos] == ',') {
            cont = 1;
            break;
        }
        if(!isNum(buf[num_tail_pos])) return 0;
    }
    if(sscanf(buf, "%d", start) == EOF) return 0;
    if(!cont) {
        *end = *start + 1;
        return 1;
    }
    buf = buf + num_tail_pos + 1;
    for(; buf[num_tail_pos] != '\n'; ++num_tail_pos) {
        if(!isNum(buf[num_tail_pos])) return 0;
    }
    if(sscanf(buf, "%d", end) == EOF) return 0;
    return 1;
}

int openFile(struct buffer * buf, char * path) {
    FILE * fp = fopen(path, "r");
    char tmp[128];
    if(fp == NULL) {
        printf("Error:Cannot read %s.\n", path);
        return 0;
    }
    char * ret = fgets(tmp, sizeof(tmp), fp);
    if(!newString(tmp, strlen(tmp), buf)) return 0;
    if(ret == NULL) return 1;
    while(1) {
        ret = fgets(tmp, sizeof(tmp), fp);
        if(ret == NULL) break;
        insertString(buf, tmp, buf->len, strlen(tmp));
    }
    return 1;
}

int writeFile(struct buffer * buf, char * path) {
    FILE * fp = fopen(path, "w");
    if(fp == NULL) {
        printf("Error:Cannot open %s.\n", path);
        return 0;
    }
    if(fwrite(buf, sizeof(char), buf->len, fp) != buf->len) {
        printf("Error:Cannot write %s.\n", path);
        fclose(fp);
        return 0;
    }
    fclose(fp);
    return 1;
}

void printLines(struct buffer * buf, int start, int end) {
    int lfCount = 0;
    for(int i = 0; i < buf->len; ++i) {
        if(lfCount >= start)
            putc(buf->buf[i], stdout);
        if(buf->buf[i] == '\n') lfCount++;
        if(lfCount >= end) return;
    }
}

int appendLines(struct buffer * buf, int start, int end) {
    int lfCount = 0;
    int i = 0;
    char tmp[1024];
    for(; i < buf->len; ++i) {
        if(buf->buf[i] == '\n') lfCount++;
        if(lfCount == start+1) break;
    }
    i++;
    while(1) {
        if(fgets(tmp, sizeof(tmp), stdin) == NULL) return 1;
        if(tmp[0] == '.') return 1;
        int size = strlen(tmp);
        int retVal = insertString(buf, tmp, i, size);
        if(retVal == 0) return 0;
        i += size;
    }
    return 1;
}

void removeLines(struct buffer * buf, int start, int end) {
    int lfCount = 0;
    int st_pos = 0, en_pos = 0;
    for(int i = 0; i < buf->len; ++i) {
        if(buf->buf[i] == '\n') {
            lfCount++;
            if(start == lfCount) st_pos = i+1;
            if(end == lfCount) en_pos = i+1;
        }
    }
    if(en_pos == 0) en_pos = buf->len;
    removeString(buf, st_pos, en_pos-st_pos);
}

int main(int argc, char ** argv) {
    struct buffer textBuffer;

    if(argc == 1) {
        printf("Usage: %s [file name].", argv[0]);
        return 1;
    }
    if(!openFile(&textBuffer, argv[1])) return 1;

    while(1) {
        char tmp[128];
        printf("> ");
        fgets(tmp, sizeof(tmp), stdin);
        if(tmp[0] == '\n') {
            continue;
        }
        int start, end;
        if(!parse(tmp, &start, &end)) {
            switch(tmp[0]) {
                case 'w':
                    writeFile(&textBuffer, argv[1]);
                    break;
                case 'q':
                    exit(0);
            }
        }
        switch(tmp[0]) {
            case 'p':
                printLines(&textBuffer, start, end);
                break;
            case 'a':
                if(!appendLines(&textBuffer, start, end)) {
                    puts("malloc error.");
                    exit(1);
                }
                break;
            case 'r':
                removeLines(&textBuffer, start, end);
                break;
        }
    }
}

File: OverviewInformationTechnology/Program1/Ruby/Section1.rb
def parse(str)
    ret = /^.(\d+),(\d+)$/.match(str)
    if ret == nil then
	ret = /^.(\d+)$/.match(str)
	if ret == nil then
            return nil
        else
            return [ret.captures[0], (ret.captures[0].to_i + 1).to_s]
        end
    else
        return ret.captures
    end
end

def openFile(path)
    fp = nil
    begin
        fp = File.open(path, "r")
    rescue => e
        puts ("Error:Cannnot read " + path + ".")
        puts e
        return nil
    end
    ret = ""
    fp.read(nil, ret)
    fp.close
    return ret
end

def writeFile(str, path)
    fp = nil
    begin
        fp = File.open(path, "w")
    rescue => e
        puts ("Error:Cannnot open " + path + ".")
        puts e
        return nil
    end
    begin
        fp.write str
    rescue => e
        puts ("Error:Cannnot write " + path + ".")
        puts e
        return nil
    end
    fp.close
    return true
end

def printLines(str, start, e) # e is end
    lfCount = 0
    str.chars.each do |c|
        if lfCount >= start then
            putc c
        end
        if c == "\n" then
            lfCount = lfCount + 1
        end
        if lfCount >= e then
            return
        end
    end
end

def appendLines(str, start, e) # e is end
    lfCount = 0
    i = 0
    (0...str.length).each do |i|
        if str[i] == "\n" then
            lfCount = lfCount + 1
        end
        if lfCount == (start+1)
            break
        end
    end
    while true
        line = $stdin.gets
        if line == nil then
            return
        end
        if line[0] == "." then
            return
        end
        str.insert(i, line)
        i = i + line.length
    end
end

def removeLines(str, start, e) # e is end
    lfCount = 0
    st_pos = 0
    en_pos = 0
    i = 0
    (0...str.length).each do |i|
        if str[i] == "\n" then
            lfCount = lfCount + 1
            if start == lfCount then
                st_pos = i + 1
            end
            if e == lfCount then
                en_pos = i + 1
            end
        end
    end
    if en_pos == 0 then
        return str[0...st_pos]
    else
        return str[0...st_pos] + str[en_pos...str.length]
    end
end

# Main Function
if ARGV.length == 0 then
    puts "Usage: <file name>."
    exit 1
end
text = openFile(ARGV[0]) 
if text == nil then
    exit 1
end

while true
    print "> "
    tmp = $stdin.gets()
    if tmp[0] == "\n" then
        next
    end
    parsed = parse(tmp)
    if parsed == nil then
        case tmp[0]
        when "w" then
            writeFile(text, ARGV[0])
        when "q" then
            exit 0
        end
    else
        case tmp[0]
        when "p" then
            printLines(text, parsed[0].to_i, parsed[1].to_i)
        when "a" then
            appendLines(text, parsed[0].to_i, parsed[1].to_i)
        when "r" then
            text = removeLines(text, parsed[0].to_i, parsed[1].to_i)
        end
    end
end